GVFS/GVFS.Common/GVFSEnlistmentStatistics.cs

```diff
@@ -0,0 +1,189 @@
+﻿using GVFS.Common.Database;
+using System;
+using System.Collections.Generic;
+using System.IO;
+
+namespace GVFS.Common
+{
+    public class GVFSEnlistmentStatistics
+    {
+        private List<string> gitPaths;
+        private List<string> placeholderFolderPaths;
+        private List<string> placeholderFilePaths;
+        private List<string> modifiedFolderPaths;
+        private List<string> modifiedFilePaths;
+        private int trackedFilesCount;
+        private int placeholderCount;
+        private int modifiedPathsCount;
+        private Dictionary<string, int> trackedFilesDirectoryTracking;
+        private Dictionary<string, int> hydratedFilesDirectoryTracking;
+
+        public GVFSEnlistmentStatistics(List<string> gitPaths, List<string> placeholderFolderPaths, List<string> placeholderFilePaths, List<string> modifiedFolderPaths, List<string> modifiedFilePaths)
+        {
+            this.gitPaths = gitPaths;
+            this.placeholderFolderPaths = placeholderFolderPaths;
+            this.placeholderFilePaths = placeholderFilePaths;
+            this.modifiedFolderPaths = modifiedFolderPaths;
+            this.modifiedFilePaths = modifiedFilePaths;
+        }
+
+        public void CalculateStatistics(string parentDirectory)

```

Comment:
Yes, we calculate statistics based on all of the paths underneath the specified one

Reply:
Have we considered returning a statistics object (i.e DirectoryStatisticsResult ) versus holding the results in this class?  @jamil for his thoughts there

Reply:
Yeah I think that is a good idea because as it is the `parentDirectory` is never saved and the results that get set are not tied to it so when you call to get any of the information about statistics you have to know what the `parentDirectory` was that got passed to `CalculateStatistics`.  So maybe this class should be a `GVFSEnlistmentStatisticsCalculator` and the `Calculate` would return the `DirectoryStatisticsResult` that would have the result for the directory with the name of the directory.

Reply:
@kewillford that was my thought

If we are calculating statistics for a directory does absoluteDirectory or relativeDirectory work?  I'm not sure if we need `parent`

Reply:
All directories are relative to the `src` folder of the enlistment. That's the most standard format paths are given in. I don't think that it makes sense to have a calculator and result object distinction. I think that the focus of the project is to provide an interactive back end for the CLI and any other applications that come out of it, it isn't meant to be passed around contexts. To move the data around, there will eventually be a `--json` option which will use a json schema to actually pass around the results you were describing if it needed to be moved, and at that time it might make sense to build out a separate object for the result, but as it stands I think the current design is the least convoluted and most logical way of accomplishing what was detailed in the design document.

Reply:
> Yeah I think that is a good idea because as it is the parentDirectory is never saved and the results that get set are not tied to it

> I don't think that it makes sense to have a calculator and result object distinction.

Tieing the state of the class to the last value passed into `calculate` makes it easy for another component to come in and calculate data for a new directory without everything in the system being aware of this change.

The way I understand it - this class takes in "persistent" data via the constructor. This is the data that doesn't change for a given setup and can be re-used to calculate statistics for multiple directories. The calculate method then calculates the statistics for a given directory.  As the lifetime / usage of this data are different (persistent and re-used, vs changing for each directory), can we separate the handling of them?